pkg/xds/generator/egress/internal_services_generator.go

```diff
@@ -0,0 +1,204 @@
+package egress
+
+import (
+	"sort"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+	envoy_common "github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_clusters "github.com/kumahq/kuma/pkg/xds/envoy/clusters"
+	envoy_endpoints "github.com/kumahq/kuma/pkg/xds/envoy/endpoints"
+	envoy_listeners "github.com/kumahq/kuma/pkg/xds/envoy/listeners"
+	"github.com/kumahq/kuma/pkg/xds/envoy/tls"
+)
+
+type InternalServicesGenerator struct {
+}
+
+// Generate will generate envoy resources for one, provided in ResourceInfo mesh
+func (g *InternalServicesGenerator) Generate(
+	_ xds_context.Context,
+	info *ResourceInfo,
+) (*core_xds.ResourceSet, error) {
+	resources := core_xds.NewResourceSet()
+
+	apiVersion := info.Proxy.APIVersion
+	endpointMap := info.EndpointMap
+	destinations := g.buildDestinations(info.TrafficRoutes)
+	services := g.buildServices(endpointMap)
+
+	g.addFilterChains(apiVersion, destinations, endpointMap, info)
+
+	cds, err := g.generateCDS(apiVersion, services, destinations)
+	if err != nil {
+		return nil, err
+	}
+	resources.Add(cds...)
+
+	eds, err := g.generateEDS(apiVersion, services, endpointMap)
+	if err != nil {
+		return nil, err
+	}
+	resources.Add(eds...)
+
+	return resources, nil
+}
+
+func (*InternalServicesGenerator) generateEDS(
+	apiVersion envoy_common.APIVersion,
+	services []string,
+	endpointMap core_xds.EndpointMap,
+) ([]*core_xds.Resource, error) {
+	var resources []*core_xds.Resource
+
+	for _, serviceName := range services {
+		endpoints := endpointMap[serviceName]
+
+		cla, err := envoy_endpoints.CreateClusterLoadAssignment(serviceName, endpoints, apiVersion)
+		if err != nil {
+			return nil, err
+		}
+
+		resources = append(resources, &core_xds.Resource{
+			Name:     serviceName,
+			Origin:   OriginEgress,
+			Resource: cla,
+		})
+	}
+
+	return resources, nil
+}
+
+func (*InternalServicesGenerator) generateCDS(
+	apiVersion envoy_common.APIVersion,
+	services []string,
+	destinationsPerService map[string][]envoy_common.Tags,
+) ([]*core_xds.Resource, error) {
+	var resources []*core_xds.Resource
+
+	for _, serviceName := range services {
+		tagSlice := envoy_common.TagsSlice(append(destinationsPerService[serviceName], destinationsPerService[mesh_proto.MatchAllTag]...))
+
+		tagKeySlice := tagSlice.ToTagKeysSlice().Transform(envoy_common.Without(mesh_proto.ServiceTag), envoy_common.With("mesh"))
+
+		edsCluster, err := envoy_clusters.NewClusterBuilder(apiVersion).
+			Configure(envoy_clusters.EdsCluster(serviceName)).
+			Configure(envoy_clusters.LbSubset(tagKeySlice)).
+			Configure(envoy_clusters.DefaultTimeout()).
+			Build()
+
+		if err != nil {
+			return nil, err
+		}
+
+		resources = append(resources, &core_xds.Resource{
+			Name:     serviceName,
```

Comment:
The naming of the clusters (and endpoints) is a problem.

Because we have a generator for every mesh, if there is the same service in different meshes, the cluster will be overridden. It works in zone ingress generator, because it's for all meshes and then service from proper mesh is picked by `mesh` tags using LbSubset.

Reply:
addressed